modoboa_amavis/utils.py

```diff
@@ -0,0 +1,134 @@
+
+
+def smart_text(value, *args, **kwargs):
+    if isinstance(value, memoryview):
+        value = value.tobytes()
+    return django_smart_text(value, *args, **kwargs)
+
+
+def fix_utf8_encoding(value):
+    """Fix utf-8 strings that contain utf-8 escaped characters.
+
+    msgs.from_addr and msgs.subject potentialy contain badly escaped utf-8
+    characters, this utility function fixes that and should be used anytime
+    these fields are accesses.
+
+    Didn't even know the raw_unicode_escape encoding existed :)
+    https://docs.python.org/2/library/codecs.html?highlight=raw_unicode_escape#python-specific-encodings
+    https://docs.python.org/3/library/codecs.html?highlight=raw_unicode_escape#python-specific-encodings
+    """
+    assert isinstance(value, six.text_type), \
+        ("value should be of type %s" % six.text_type.__name__)
+    return value.encode("raw_unicode_escape").decode("utf-8")
+
+
+class ConvertFrom(Func):
+    """Convert a binary value to a string.
+    Calls the database specific function to convert a binary value to a string
+    using the encoding set in AMAVIS_DEFAULT_DATABASE_ENCODING.
+    """
+
+    """PostgreSQL implementation.
+    See https://www.postgresql.org/docs/9.3/static/functions-string.html#FUNCTIONS-STRING-OTHER"""
+    function = "convert_from"
+    arity = 1
+    template = "%(function)s(%(expressions)s, '{}')".format(
+        settings.AMAVIS_DEFAULT_DATABASE_ENCODING)
+
+    def as_mysql(self, compiler, connection):
+        """MySQL implementation.
+        See https://dev.mysql.com/doc/refman/5.5/en/cast-functions.html#function_convert"""
+        return super(ConvertFrom, self).as_sql(
+            compiler, connection,
+            function="CONVERT",
+            template="%(function)s(%(expressions)s USING {})".format(
+                settings.AMAVIS_DEFAULT_DATABASE_ENCODING),
+            arity=1,
+        )
+
+    def as_sqlite(self, compiler, connection):
+        """SQLite implementation.
+        SQLite has no equivilant function, just return the field."""
+        return super(ConvertFrom, self).as_sql(
+            compiler, connection,
+            template="%(expressions)s",
+            arity=1,
+        )
+
+
+def check_amavis_database_encoding():
+    """Check AMAVIS_DEFAULT_DATABASE_ENCODING has been set to the correct value."""
+    db_engine = settings.DATABASES["amavis"]["ENGINE"]
+    sql_query = None
+    if "postgresql" in db_engine:
+        sql_query = "SELECT pg_encoding_to_char(encoding) "\
+            "FROM pg_database "\
+            "WHERE datname = %s;"
+    elif "mysql" in db_engine:
+        sql_query = "SELECT DEFAULT_CHARACTER_SET_NAME "\
+            "FROM INFORMATION_SCHEMA.SCHEMATA "\
+            "WHERE SCHEMA_NAME = %s;"
+    elif "sqlite" in db_engine:
+        sql_query = "PRAGMA encoding;"
+    else:
+        warnings.warn(
+            _("Modoboa Amavis has not been tested using a %s database")
+            % db_engine)
```

Comment:
Add a return in the else and remove the test on sql_query ?

Reply:
fixed

Reply:
I might have a better solution for `check_amavis_database_encoding()` using [Djangos system check framwork](https://docs.djangoproject.com/en/1.11/topics/checks/) but I need to look into it and do some testing.

If it works out we update the docs to say run `manage.py check --deploy` after each upgrade and these checks can catch thinks like `AMAVIS_DEFAULT_DATABASE_ENCODING` not being set correctly.

Reply:
Indeed, it seems to be a better option.

Reply:
implemented 15d9cc9